TypeDB Clients documentation technical review (#679)
graknlabs/docs: 02-clients/01-studio.md

---
pageTitle: TypeDB Studio
keywords: typedb, client, studio
longTailKeywords: typedb studio, typedb studio preferences, typedb studio connection
summary: Brief intro for TypeDB Studio.
toc: false
---

> # TypeDB Studio

Reviewer: Would a hero image before everything else be gratuitous? Something like the image from the README:
![image](https://user-images.githubusercontent.com/51956016/235954737-74961ed2-7b23-4891-a693-afcfe3828f02.png)

Reviewer: TypeDB Studio is inherently a graphical thing, so it makes sense to start with an image. Though, perhaps an image that shows every feature in use could be overwhelming at the beginning of the tutorial.

Author: I like the idea in general. Even tho we don't want the picture to be too big. I'll think about a balanced solution here.

Author: Fixed. Also the image I added is not as cool because I had to use IAM schema and micro-dataset and they don't have that sweet data clusterization.

TypeDB Studio is an IDE specifically designed to work with 

> [TypeQL]() query language and TypeDB databases.

Reviewer: The QL means query language already ;)

Author: But the name of the query language is still `TypeQL`, isn't it? Like the name of the most known query language is `SQL`. That is why I suggest that the line is correct (also, missing a link for the term, because we don't have that section yet).

Reviewer: Could we go with "[...] work with TypeQL (TypeDB's query language) and TypeDB databases."?

Author: We can't risk it being "TypeQL and TypeDB databases".

Updated the definition in https://github.com/vaticle/docs/pull/679/commits/8fff3059fd2179df21e5117685910bf9084e84e4
Please check

Reviewer: agreed, resolved.

It has unique set of functions and abilities to help with development using TypeDB:

- Friendly graphical user interface (GUI) to design schemas and queries for TypeDB databases.
- Syntax highlighting for TypeQL.
- Connection to TypeDB servers, including handling transactions.
- Types explorer for a TypeDB database.
- Query execution, including schema and data, read and write transactions.
- Local query validation with constraints of an actual database schema before sending the query to server.
- Graph visualization for query responses.

## Installation

### Requirements

TypeDB Studio do not require any additional packages to be installed prior the installation, but it needs a TypeDB 
server to perform most it's functions. We can use either remote or local TypeDB server.

<div class="note">
[Note]
The default TypeDB settings are to run server listening for `localhost` connections with port number `1729`.
</div>

<div class="note">
[Important]
We recommend using the latest version of TypeDB server and a [compatible](#version-compatibility) version of TypeDB 
Studio. 
</div>

### Download & install

Download TypeDB distributive/package for any popular OS family from the

> [Download page](https://vaticle.com/download#typedb-studio. 

Reviewer: ```suggestion
[Download page](https://vaticle.com/download#typedb-studio). 
```

Follow proper installation procedure of the OS in use.

<div class="note">
[Important]
Make sure to grant access for installation wizard to install TypeDB Studio in case any security prompts will appear.
</div>

<div class="note">
[Note]
For **macOS** TypeDB Studio can be installed via homebrew:

```
brew tap vaticle/tap
brew install --cask vaticle/tap/typedb-studio
```
</div>

## Prepare for work

Right after starting TypeDB Studio we can see its main window as it is shown below.

> ![TypeDB Studio GUI](../images/studio/studio-gui.png)

Reviewer: I think this image could use a little work. Some suggestions:
* It would be great if we used our default font (Titillium Web) for the numbers: https://fonts.google.com/specimen/Titillium+Web
* We should try putting boxes around the sections of Studio we're highlighting, preferably with rounded corners.
* Let's use our light green accent (#02DAC9) for the numbers and the boxes rather than the current red.

Reviewer: Likewise for the rest of the places we've added highlights (arrows, boxes, circles, etc).

Author: Implemented with the font, but the Pink color from our new color palette. Haikal chose Pink.

Reviewer: I love it! Looks much better.

Enter the name for the new database at the text field at the bottom of the `Manage Databases` window and hit 
[**Create**].

![Database Manager Empty](../images/studio/databases-interface-no-databases.png)

Our database appeared at the list of databases. Now we can close the `Manage Databases` window by clicking the 
[**Close**] button.

![Database Manager with IAM database](../images/studio/databases-interface-iam-database.png)

### Select active database

TypeDB Studio lets us work with one database at a time. Thus, to perform any query we need to choose a database 
first. 

We can select it by clicking the dropdown menu at the toolbar titled [**database (none)**] or with a name of 
previously selected database. It is the only dropdown menu in the toolbar and is located immediately right of the 
databases icon.

<div class="note">
[Note]
If the dropdown menu isn't active it usually means that we didn't connect to a server yet.
</div>

After selecting the database to work with from a list of all databases on the server Studio connected with, we will 
see types hierarchy of the selected database in the Types browser (bottom left of the screen). For example, for any 

> empty database (with no schema defined) it will show only the base types:

Reviewer: A picture of this would be nice to help the user see they're on the right track.

Author: This is not the place to talk about the Types explorer in detail. It's merely a section on how to select a database.

TypeDB databases. For that we have the Text editor panel with TypeQL syntax highlighting and the Toolbar with 
transaction controls.

#### Text editor panel

Text editor panel doesn't have a visible title, but it's the biggest area in the middle of the TypeDB Studio that is 
being used to work with text information. It does have tabs at the top to switch between different opened files.

The currently selected file's content displayed in the main section with lines numeration at the left edge of the 
Text editor panel.

All text in the Text editor panel will be displayed with TypeQL syntax highlighting.

In the bottom right corner of TypeDB Studio window there is an indicator for cursor location that contains the line 
number and current symbol position in that line.

The following functions can be activated from the context menu available by right-clicking inside the Text editor 
main section:

- Copy/Cut/Paste
- Find/Replace
- Run File/Run Selection
- Increase/Decrease/Reset Text Size

#### Run panel

Under the Text editor panel there is the Run panel, but it is collapsed by default. To expand it perform any query or 

> just click on the reversed tick icon in the bottom right corner of TypeDB Studio window, above the cursor location 

Reviewer: ```suggestion
just click on the upwards-pointing chevron in the bottom right corner of TypeDB Studio window, above the cursor location 
```

Reviewer: or 'up arrow'.

indicator.

Run panel has 2 sets tabs:

- Tabs at the top of the Run panel are for different query attempts. 
  
  By default, every query sent will overwrite the 
  existing tab, but if we enable the pin icon at the beginning of the tab name by clicking it — the tab will become 
  pinned and stay. A new tab will appear on the right from the last pinned one upon next query execution.

- Tabs at the bottom of the Run panel are for different results for the currently selected query attempt (tab at the 
  top of the Run panel). These include the `Log` tab with direct textual Studio output and the `Graph` tabs (one per 
  every query statement in the Text editor panel) with graph visualization of response.

#### Transaction control

To execute a query written in the Text editor panel we have the **Toolbar** elements located right from the database 
selection drop-down menu. See below.

> ![Transaction controls](../images/studio/studio-transaction-controls.png)

Reviewer: Let's take this picture again with the database name as 'iam' rather than 'iam-4'.

Author: I did that and copy-pasted all buttons in their active (not greyed out) state.

From left to right:

- [Session](../0001-typedb/02-dev/01-connect.md#sessions) type switch:
  - `schema` — to use schema session type
  - `data` — to use data session type
- [Transaction](../0001-typedb/02-dev/01-connect.md#transactions) type switch:
  - `write` — to use write transaction
  - `read` — to use read transaction
- [**snapshot**] button — enables snapshot feature that encapsulates transaction in a snapshot of a database's data 

>   until the transaction committed or closed. Write transactions are always snapshoted. Only available for read 
>   transactions.

Reviewer: ```suggestion
  until the transaction committed or closed. Write transactions are always snapshoted. Always enabled for write transactions, may be disabled for read 
  transactions.
```

Author: ```suggestion
  until the transaction committed or closed. Write transactions are always snapshoted. May be enabled or disabled for read 
  transactions.
```

- [**infer**] button — enable [inferring data](../0001-typedb/02-dev/06-infer.md) by rules. Only available for read 
  transactions.

> - [**explain**] button — enable [explanations](../0001-typedb/02-dev/06-infer.md#explain-query) mechanics by providing 

Reviewer: \+ "Only available for read transactions."

>   `explainables` methods for results.

Author: ```suggestion
- [**explain**] button — enable [explanations](../0001-typedb/02-dev/06-infer.md#explain-query) mechanics by providing 
  `explainables` methods for results. Only available for read transactions with **snapshot** and **infer** enabled.
```

- [**Transaction status**] (Circle (●) icon) indicator — transaction activity indicator. Green if there is a 
  transaction 

>   opened by TypeDB Studio at this moment. Grey if there is no opened transaction. Rotating animation — TypeDB 

Reviewer: ```suggestion
  opened by TypeDB Studio at this moment. Gray if there is no opened transaction. Rotating animation — TypeDB 
```
We have to speak Amer*can now :(

Run panel.

If a `write` transaction committed successfully — a blue pop-up with a transaction confirmation is displayed. See 
example below.

![Commit the Transaction](../images/studio/project-schema-committed.png)

## Types explorer capabilities

Under the Project panel that shows the Project directory TypeDB has Types panel that shows all types of the schema of 
the current database. 

<div class="note">
[Note]
If the Types panel is not displayed — click the Types vertical tab on the left edge of the TypeDB Studio window.
</div>

All types are divided in three top level [base types](../0001-typedb/02-dev/02-schema.md#types):

- Attribute
- Entity
- Relation

![Types explorer](../images/studio/types-explorer.png)

All types displayed in the exact hierarchy they were created (subtypes inside their supertypes). We can easily 
collapse or expand all the types in the structure tree by clicking the associated buttons on the top of the Types panel 
with arrows pointing down or up respectively.

Types panel can be closed by clicking the X icon in the top right corner of the panel. To reopen the panel click the 
Types tab on the left edge of the TypeDB Studio window.

### Export schema

> We can export schema of the database by clicking the [**Export Schema**] button (second in row from left to right) on 

Reviewer: The button is labeled "Export Schema Types" to indicate that it does not include rules. The warning below is still valid as it's annoying.

Author: Was the third word of the button named added in 2.17? Because in older version of Studio (when the text was written) I do remember it wasn't there.

[more](../0001-typedb/02-dev/02-schema.md#undefine-a-type) on these conditions.
</div>

## Settings

To configure TypeDB Studio open the **Manage Preferences** window by clicking on the button with the gear (⚙) icon 
in the top right corner of the TypeDB Studio window.

TypeDB Studio has the following settings available:

- Graph Visualiser
  - Enable Graph Output — Turns on visualization of query results as graphs on a separate tab of the Run panel. 
    Default value: **On**.
- Project Manager
  - Set Ignored Paths — All paths mentioned here will be invisible in any Project directory opened in Studio. 
    Default value: `.git`.
- Query Runner
  - Default Match Query Limit — Limits the maximum number of results to be returned from a get query to a database 
    if no limit explicitly stated in the query. Default value: **1000**.
- Text Editor
  - Enable Autosave — Enables autosave function for files opened in TypeDB Studio. Default value: **On**.

## Version Compatibility

|      Studio      |      TypeDB      |  TypeDB Cluster  |
|:----------------:|:----------------:|:----------------:|
| 2.14.1 to 2.14.2 | 2.14.1 to 2.14.2 |      2.14.1      |

> |      2.11.0      |      2.11.1      | 2.11.1 to 2.11.2 |

Reviewer: We can add all of 2.17.0 here!

Author: Ideally I would prefer not to mix it up. We have added the version in to the master branch. So if we don't touch the table it should be updated automatically upon merge (not by this PR probably). But that is a big IF, considering many factors. 
Anyway - we can double-check versions in those tables as part of releasing this new docs, just not to update all the PRs all the time new release happens. I think that would be best.

Reviewer: Makes sense.

2. Choose the correct session (`schema`/`data`) type and transaction (`read`/`write`) type to run the query by  
   clicking on those options on the [**Toolbar**] switches.

3. Run query by clicking on the [**Run Query**] (▶) button. [**Run**] panel will be expanded on the bottom of the  
   query to show the results of the query execution attempt.

   ![Write the Schema](../images/studio/project-schema-query-run.png)

4. If it's the `write` transaction type — wait for transaction control options to be ready and commit or close the 
   transaction (by clicking on the [**Commit transaction**] (✔) or [**Close transaction**] (x) buttons respectively).

The result of the query is displayed in the [**Run**] panel under the Text editor.

If any error occurs during the query validation or transaction it is shown as red popup in the bottom right corner of 
the TypeDB Studio window. Additional information about the error is displayed in the results of the query in the 
[**Run**] panel.

If a `write` transaction committed successfully — a blue pop-up with a transaction confirmation is displayed. See 
example below.

![Commit the Transaction](../images/studio/project-schema-committed.png)

> ## Types explorer capabilities

Reviewer: We call this the 'Type Browser' in our code, so let's reflect this here too.

Author: But why we should stick to the existing terminology if we can improve it? 
Let's discuss what we should call it in our documentation for our users and then adapt the source codes to reflect the naming if needed.

We clearly have these explorer vs browser terms. And we have to name the Types panel (or we can leave it as the Types panel) and the tab with Type detailed information.

**Explorer** is more like exploring all different files/types/objects, moving through the hierarchy. So it's closer to the panel. 

**Browser** is more about seeing a detailed view of a singular type at a time (with the potential ability to move to another). So it's closer to the tab.

That is why I called them this way. We can skip the panel (as we can call it the Types panel) and just call the tab the Type page, the Type tab, or something like that. But that will belittle the features. As we are building an IDE, we should really have some ambition in the naming of functional elements IMHO.

Reviewer: I see your argument. I think we should create an issue specific to the nomenclature of Studio components and come to agreement on all of the naming that might be more understandable to users if changed.

In the mean time, I think it's important that we don't enter a state where the source code and the documentation diverge in how things are named. It's a tenant of our design philosophy that we have a ubiquitous language (at least per product) that we adhere to across documentation, code and discussions.

Author: 1. We have discussed the approach with Shane and I think he is more flexible with this. We are re-implementing our external communications framework and rewriting the documentation. It is time to change the terminology now if we think it will help with our external communications. The only question is what will be chosen as the right terminology.
2. There is a meeting about terminology planned. I'm waiting for Shane to approve it before proceeding.

on the top of the [**Types**] panel with double chevrons (arrow heads) pointing down or up respectively.

![Types explorer](../images/studio/types-explorer.png)

Double-clicking on any type will open the [**[Type browser](#type-browser)**] with information about the selected type 
in a tab of the [**Text editor**].

[**Types**] panel can be closed by clicking the **X** icon in the top right corner of the panel. To reopen the panel 
click the [**Types**] tab on the left edge of the TypeDB Studio window.

### Export schema

We can export schema of the database by clicking the [**Export Schema Types**] button (second in row from left to 
right) on the top of the [**Types**] panel. It creates a new tab with inserted TypeQL code that creates the exact 
schema that is in the current database.

<div class="note">
[Warning]
The [**Export Schema Types**] button doesn't include rules into its output. That will be fixed in some of the next 
versions of the TypeDB Studio. 
</div>

### Type browser

By double-clicking any type we can open the type browser window that shows all information about the selected type and
allows us to easily edit the type. 

> ![Types browser](../images/studio/types-browser.png)

Reviewer: This is a type page! The Type Browser is, as above, the panel on which all the types are listed in Studio's main view.
